### DISCUSSION TITLE
Should iterator operators forward return() and throw() to their underlying iterators?

### DISCUSSION THREAD

**Discussion by mattbishop** [vitaly-t/iter-ops#211] [Q&A] [answered]
I am working on a project that has an async iterator backed by a DB cursor. This iterator is part of a `pipeAsync` chain that looks like this:

```js
  const elements = pipeAsync(
    // firstBatch is an array of rows, read and examined to collect outerId
    firstBatch,
    concat(dbCursorIterator),
    // dbCursorIterator is a batch of 100 rows
    flat(2),
    map((row) => rowToElement(outerId, row)))

  const response = Readable.from(elements)
```

This Readable is passed back to an HTTP response stream where it pulls the rows through the cursor and sends them to the client.

On occasion, the response Readable is destroyed before the cursor is complete. The common case is when the client closes the connection early.

Readable knows it is wrapping an AsyncIterator, so it calls iter.return() or iter.throw(error), depending on the circumstances. The iter-ops operators and methods do not seem to have these methods, so Readable cannot communicate the early termination down to my DB cursor iterator. As a result, the cursor hangs open and uses up resources.

Is this a bug, a feature request, or my misunderstanding? I have looked at the error handling page but this is not a case where an iterator is throwing an error, it's an outside signal to terminate the iterator early.

**Comment by vitaly-t** [comment, marked as answer]
> Is this a bug, a feature request, or my misunderstanding? I have looked at the error handling page but this is not a case where an iterator is throwing an error, it's an outside signal to terminate the iterator early.

It is a misunderstanding. This library only handles iterable sequences, it does not handle life cycle of external objects. I understand that your stream object is released before the cursor is finished, and so the source (db cursor) remains open.

You would have to sync up those external objects by yourself, something like this:

```ts
const response = Readable.from(elements)

response.once('close', () => {
   dbCursorIterator.close(); // or whatever method it has to close your cursor
});
```

Because as far as the iterable is concerned here, nothing has happened, iteration just paused, that's it.

For reusability, you can create a helper function like this:

```ts
// creates Readable from an Iterable, and auto-closes the source db cursor;
readableForCursor(cursor, i: Iterable): Readable {
    const r = Readable.from(i);
    r.once('close', () => {
       cursor.close(); // or whatever method it has to close your cursor
    });
    return r;    
}
```

And then use it instead:

```ts
const response = readableForCursor(dbCursorIterator, elements);
```

**Comment by vitaly-t** [comment]
A better approach would be to take advantage of the new operator [consume](https://vitaly-t.github.io/iter-ops/functions/consume) that was added in version 2.9.0, and create your own custom operator (see examples in [iter-ops-extra](https://github.com/vitaly-t/iter-ops-extras)):

```ts
import {consume, Operation} from 'iter-ops';
import {Readable} from 'stream';

// safe Readable creator, with auto-close of cursor-parameter;
export function streamCursor<T>(cursor: CursorType): Operation<T, Readable> {
    return source => consume(i => {
        const r = Readable.from(i);
        r.once('close', () => {
            cursor.close(); // or whatever the method is
        });
        return r;
    })(source);
}
```

Then your code can be simplified to this:

```ts
 const elements = pipeAsync(
    // firstBatch is an array of rows, read and examined to collect outerId
    firstBatch,
    concat(dbCursorIterator),
    // dbCursorIterator is a batch of 100 rows
    flat(2),
    map((row) => rowToElement(outerId, row)),
    streamCursor(dbCursor) // or whatever cursor object that you have
);

const response = await elements.first; // safe Readable
```

@mattbishop

**Comment by mattbishop** [comment]
Thank you for taking the time to respond so thoroughly! I wasn't aware of the custom operators. I can think of ways of using that on another project.

 I added a "close" listener as you suggested and now I can clean up the cursor nicely in all cases, not just errors.

**Comment by vitaly-t** [comment]
Here's another version of a reusable operator for producing a `Readable`, you can just put it into a separate file ;)

```ts
import {consume, Operation} from 'iter-ops';
import {Readable} from 'stream';

export type ReadableEvent = 'close' | 'data' | 'end' | 'error' | 'pause' | 'readable' | 'resume';

const readableEvents: ReadableEvent[] = ['close', 'data', 'end', 'error', 'pause', 'readable', 'resume'];

export function toReadable<T>(cb?: (event: ReadableEvent, param?: any) => void): Operation<T, Readable> {
    return source => consume(i => {
        const r = Readable.from(i);
        if (typeof cb === 'function') {
            readableEvents.forEach(e => r.once(e, param => cb(e, param)));
        }
        return r;
    })(source);
}
````

In your case, you would use it like this:

```ts
const i = pipe(
    inputCursor,
    // other operators
    toReadable(event => {
        if(event === 'close') {
            inputCursor.close();
        }
    })
); //=> AsyncItreableExt<Readable>
```

**Comment by vitaly-t** [comment]
Added operator [toReadable](https://github.com/vitaly-t/iter-ops-extras/blob/main/src/to-readable.ts) within [iter-ops-extras](https://github.com/vitaly-t/iter-ops-extras), which can be used like this:

```ts
toReadable(r => {
   r.once('close', () => {
       // stream closed
   })
})
```

**Reply by mattbishop** [reply]
Thanks Vitaly.

Going back to the original question, does it make sense for AsyncIterator wrappers created by iter-ops to forward `return()` and `throw()` calls down to their enclosed iterables? I understand they only make sense in AsynIterators.

The reason I ask is `toReadable` is good when one has access to the underlying data source, but that may not always be the case.

**Reply by vitaly-t** [reply]
Are you asking about a specific operator? They behave differently, depending on operator logic. As for `throw`, it is passed on through iterators, so it can be handled in a generic way, via [catchError](https://vitaly-t.github.io/iter-ops/functions/catchError.html). It works that way for both sync and async iterables.

You do not need to use `toReadable`, if you don't want to, see example for the [compose](https://vitaly-t.github.io/iter-ops/functions/consume.html) operator, or you can create it outside iteration as you did before.

**Reply by mattbishop** [reply]
I mean generally. These two methods on AsyncIterator are optional, but valuable when present. I know Readable will call them when stream ends early or an error occurs. It would be good if every iter-ops Async wrapper would forward the call down so it can be handled when called.

**Reply by mattbishop** [reply]
For instance, in this wrapper:

https://github.com/vitaly-t/iter-ops/blob/main/src/helpers.ts#L53

The library could add these methods:

```js
    return(): { it.return && it.return(); },
    throw(err): { it.throw && it.throw(err); }
```

**Reply by vitaly-t** [reply]
You are confusing iterators with generators. What you refer to is part of the generators spec. This library works with iterables and iterators, not with generators that produce iterables.

**Reply by mattbishop** [reply]
OK, regardless. In Node, Readable.close() will check its wrapped `iterator` instance for for both `throw` and `return` methods and call them given the state of the close() operation.

https://github.com/nodejs/node/blob/ba67fe66eb7777d5055c785be153374843fc647e/lib/internal/streams/from.js#L65

What I am asking is that iter-ops wrappers accept those calls and forward them to the wrapped iterator so they can dispose of their own resources appropriately when the Reader is closed.

**Reply by vitaly-t** [reply]
Again, those are extensions for generators, they are not part of the iterable spec, and therefore, outside the scope here. This library handles errors in the way consistent with the iterable spec, passing them downstream, nothing more nothing less.
